### PR TITLE
enhancement(preview): add blocks support

### PR DESCRIPTION
**Pull request by mathildeleg**
### What does it do?
Adds inline-editing support for Blocks fields in the live preview popover.
- Content source maps (content-source-maps.ts): extends stega encoding to the Blocks AST — walks each top-level block and encodes the first text leaf with a fieldPath marker (instead of the per-leaf path used for string fields). This lets the preview script identify blocks fields as a single editable unit.
- Preview script (previewScript.js): handles blocks fields specifically — collapses all stega spans belonging to the same blocks field into one highlight group whose bounding rect covers the entire rendered output (including container padding), detects which top-level block was double-clicked (by index), and forwards live value updates via a `strapiFieldChange` custom event instead of patching the DOM directly (blocks are rendered by the host's BlocksRenderer, not as plain text nodes).
- InputPopover: passes blockIndex from the double-click payload into context; closes on an iframe click event (replaces the previous full-iframe overlay div); adds `collisionPadding` and `maxHeight` so the popover doesn't overflow the viewport when the blocks editor is tall.
- BlocksEditor / BlocksInput / BlocksContent: new livePreviewSync prop bypasses the 300 ms debounce and syncs on every keystroke so the preview updates in real time; new blockIndex prop positions the Slate cursor at the clicked block and scrolls it into view on open; onFocus / onBlur are forwarded to BlocksContent so the popover can track editing state.
- EditorToolbarObserver: fixes a More-menu inconsistency in the popover — the BlocksDropdown label width varies with the selected block type, narrowing the toolbar container; replaced a one-shot `useLayoutEffect` measurement with a `ResizeObserver` that re-measures on every container resize, so the overflow state is always accurate regardless of which block is focused.
- Preview.tsx: posts a `STRAPI_RESCAN_HIGHLIGHTS` message to the iframe when the popover closes, so stega element groups are rebuilt after the edited content re-renders (heights change, new nodes appear).

### Why is it needed?
The Blocks rich-text field type was the last major field type unsupported by the live preview inline-editing feature. Users could hover over blocks content in the preview and see the highlight, but double-clicking had no effect. This PR closes that gap: the full editing workflow (hover → double-click → edit → live update → close) now works for Blocks fields.

### How to test it?
Requirements:
- An app with the preview configured (e.g. examples/getstarted or ask AI to create an examples/launchpad from the Launchpad repo)
- Have the Next.js server from Launchpad repo on, and server from your example app on
- An article content type with a blocks field

Steps:
  1. Open a content entry that has a Blocks field in the Edit View and navigate to the Preview tab.
  2. Hover over a paragraph, heading, or list in the preview — a purple highlight border should appear around the entire blocks field.
  3. Double-click on a specific block — the inline editing popover should open with the Blocks editor, with the cursor positioned at the block you clicked.
  4. Type something — the preview should update in real time (no 300 ms delay).
  5. Click outside the popover (anywhere in the iframe) — the popover should close and the preview should rescan highlights.
  6. Verify the More menu in the popover's Blocks toolbar behaves consistently regardless of which block type is selected (paragraph, heading, numbered  list, etc.).
  7. Test with the popover near the bottom edge of the viewport — it should flip upward and respect the available height without overflowing.

### CODE REVIEW

File: packages/core/content-manager/admin/src/preview/utils/tests/previewScript.test.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    document.body.innerHTML = '';
+    window.STRAPI_DISABLE_STEGA_DECODING = true;
+    const scriptText = fs.readFileSync(serverScriptPath, 'utf-8');
+    const config = JSON.stringify({
+      colors: PREVIEW_HIGHLIGHT_COLORS,
+      events: INTERNAL_EVENTS,
+      parentOrigin: window.location.origin,
+    });
+    // eslint-disable-next-line no-eval
+    eval(`(function() { ${scriptText}; previewScript(${config}); })()`);
```
[the remark this answers is not in the document]
Comment (mathildeleg) [reply packages/core/content-manager/admin/src/preview/utils/tests/previewScript.test.ts]: Fixed

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js [hunk 2]
```diff
@@ -82,6 +82,57 @@ function previewScript(config) {
...
+    params.delete('path');
+    return params.toString();
+  };
+
+  /**
+   * @param {unknown} value
+   * @returns {boolean}
+   */
+  const isBlocksValue = (value) => {
+    if (!Array.isArray(value) || value.length === 0) return false;
```
Comment (innerdvations) [inline packages/core/content-manager/server/src/preview/controllers/previewScript.js:110]: **Blocker — clearing blocks skips the CustomEvent path.**

`isBlocksValue` returns false for `null` and `[]`. When the editor clears a blocks field, the admin posts that value as a `strapiFieldChange`; this gate then fails and the handler falls through to `getElementsByPath` / DOM `textContent` patching on the stega spans instead of `forwardBlocksFieldChange`.

Please detect blocks via field metadata (`type=blocks` / `fieldPath` on the matched source) so empty clears still call `forwardBlocksFieldChange(field, value)` (including `null`/`[]`), and add a regression for that path.

_AI-assisted comment._

File: packages/core/content-manager/admin/src/preview/utils/previewScript.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * looks like a blocks AST, it re-dispatches the payload as a CustomEvent so
+ * host-app components (e.g. BlocksRenderer) can subscribe and re-render
+ * without the script patching the DOM directly.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function previewScript(_config: unknown): void {
+  // Inlined so the constant survives toString()+eval() with no closure.
+  const STRAPI_FIELD_CHANGE = 'strapiFieldChange';
+
+  const isBlocksValue = (value: unknown): boolean => {
```
Comment (innerdvations) [inline packages/core/content-manager/admin/src/preview/utils/previewScript.ts]: **Blocker — tests don’t cover the real preview script.**

This admin stub (+ its unit tests) never runs in the iframe. The production path is `server/.../previewScript.js`, including the null/`[]` clear behaviour above. Prefer dropping or clearly marking this as a non-SUT stub and adding tests against the real script/helpers (with a null/empty blocks regression).

_AI-assisted comment._

File: docs/docs/docs/01-core/content-manager/06-preview.md [hunk 4]
```diff
@@ -96,3 +100,78 @@ Users can configure the preview behavior from their frontend via `window` global
...
+### Stega encoding for blocks
+
+The server-side content-source-maps service (`packages/core/core/src/services/content-source-maps.ts`) has a dedicated `encodeBlocks` pass over the blocks AST. Instead of encoding every text node, it injects **one stega marker per visual block**:
+
+- `paragraph`, `heading`, and `quote` nodes: the first `{ type: 'text' }` leaf in the subtree.
+- `list` nodes: the first text leaf of each list item (recursively for nested lists).
+- `image` nodes: the `url` and `alternativeText` string fields.
+- `code` blocks are **skipped** — encoding their content would corrupt the syntax.
+
+All markers within one blocks field share the same `fieldPath` metadata key (e.g., `fieldPath=content`), which distinguishes them from regular string-field markers. The individual `path` key still varies per leaf (each leaf has a unique AST position), but the preview script uses `fieldPath` as the group key to cluster all markers into one highlight.
```
Comment (innerdvations) [inline docs/docs/docs/01-core/content-manager/06-preview.md]: **Blocker — docs don’t match encoding.**

This says the individual `path` key “still varies per leaf”, but `blocksFieldMetadata` sets `path: metadata.fieldPath` for every marker (`content-source-maps.ts`), and the unit tests assert `path=${FIELD_PATH}` alongside `fieldPath`. Please correct the contract so docs match what ships.

_AI-assisted comment._

**Review by innerdvations — CHANGES_REQUESTED** [review]
Thanks for the blocks live-preview work — direction looks right. A few issues before merge:

1. **Clearing a blocks field** sends `null`/`[]`, which `isBlocksValue` rejects, so the script falls through to DOM text patching instead of `forwardBlocksFieldChange` / `strapiFieldChange`.
2. **Tests** exercise an admin stub, not the real iframe `previewScript.js` (including the null/empty path).
3. **Docs** say each leaf’s `path` varies; encoding sets `path === fieldPath` for every marker.

Details inline.

_AI-assisted comment._

**Comment by mathildeleg** [comment]
> Thanks for the blocks live-preview work — direction looks right. A few issues before merge:
> 
> 1. **Clearing a blocks field** sends `null`/`[]`, which `isBlocksValue` rejects, so the script falls through to DOM text patching instead of `forwardBlocksFieldChange` / `strapiFieldChange`.
> 2. **Tests** exercise an admin stub, not the real iframe `previewScript.js` (including the null/empty path).
> 3. **Docs** say each leaf’s `path` varies; encoding sets `path === fieldPath` for every marker.
> 
> Details inline.
> 
> _AI-assisted comment._

Thanks for the detailed review! Here's what changed:

1. Clearing a blocks field falling through to DOM patching
Added a type field to the `strapiFieldChange` payload so the preview script can identify blocks fields even when the value is null or []. The check in `handleMessage` is now `if (type === 'blocks' || isBlocksValue(value))`, which correctly routes clears to `forwardBlocksFieldChange` instead of the DOM patching path.

2. Tests covering a stub instead of the real script
Removed the admin-side stub (`previewScript.ts`) and its test file entirely. Replaced them with a test that loads the actual `previewScript.js` source via `fs.readFileSync` and evaluates it using new Function in a jsdom environment — this avoids Istanbul's cov_* instrumentation issue while testing the real code. The `isBlocksValue.ts` extracted file was also removed; the logic stays inlined in `previewScript.js` (required by the self-contained iframe constraint) and the edge case tests live directly in the new `previewScript.test.ts.`

3. Docs incorrectly describing path per leaf
Fixed — the docs now correctly state that all markers within a blocks field share the same path value (equal to fieldPath), and that fieldPath's presence is what distinguishes blocks markers from regular string-field markers.

Bonus fix: Also fixed a cursor positioning regression where double-clicking a block in the preview opened the editor at the wrong block. The `findBlockIndex` function was using nearest-common-ancestor across all stega spans, which landed on a high-level DOM ancestor when the page had complex layout — returning an index like 79 for a field with 28 blocks. It now walks up from the clicked anchor directly, which reliably finds the actual blocks field container.

File: packages/core/core/src/services/content-source-maps.ts [hunk 5]
```diff
@@ -46,15 +48,151 @@ interface EncodingInfo {
...
+): Record<string, unknown> => {
+  if (!isObject(imageNode.image)) {
+    return imageNode;
+  }
+
+  const fieldMetadata = blocksFieldMetadata(metadata);
+  const image = { ...imageNode.image };
+
+  if (typeof image.url === 'string') {
+    image.url = encodeField(image.url, fieldMetadata);
```
Comment (giu1io) [inline packages/core/core/src/services/content-source-maps.ts]: **Blocker — after adding an image block, clicks anywhere in the preview open the blocks editor.**

https://github.com/user-attachments/assets/1ced2f23-444f-4f58-b460-f165ca422028

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js [hunk 6]
```diff
@@ -505,10 +577,148 @@ function previewScript(config) {
...
+     *
+     * @param {HighlightGroup} group
+     * @returns {HTMLElement | null}
+     */
+    const findBlocksContainer = (group) => {
+      const firstEl = group.elements.values().next().value;
+      if (!firstEl) return null;
+      let el = firstEl.parentElement;
+      while (el && el !== document.body && el !== document.documentElement) {
+        if (Array.from(el.children).some((c) => BLOCK_LEVEL_TAGS.includes(c.tagName))) {
```
Comment (giu1io) [inline packages/core/content-manager/server/src/preview/controllers/previewScript.js]: **Nested lists resolve the wrong field container.**

`BLOCK_LEVEL_TAGS` includes `OL` and `UL`, while a Strapi list can contain another list as a direct child. Since `encodeListBlock` recurses into nested `list` children, a nested list's marker sits inside the outer `<ul>` — whose direct children include the inner `<ul>` — so this loop stops at the outer list instead of the blocks field wrapper.

The highlight then covers only that list, and `findBlockIndex` (which has the identical loop at `:671`) returns an index within the list rather than the top-level Slate index. Double-clicking a nested list after a paragraph would open the editor at the wrong block.

This is the same class of problem the removed NCA code was trying to handle — its docstring specifically mentions the `<li>` case.

Please make the walk-up skip an ancestor that is itself a block-level element (or track the container from the marker's own top-level block), and add a regression with a paragraph followed by a nested list.

_Generated by Claude, edited by me._

File: docs/docs/docs/01-core/content-manager/06-preview.md [hunk 7]
```diff
@@ -96,3 +100,78 @@ Users can configure the preview behavior from their frontend via `window` global
...
+// postMessage (received in the iframe from the admin parent)
+window.addEventListener('message', (e) => {
+  if (e.data?.type === 'strapiFieldChange') {
+    const { field, value } = e.data.payload;
+    // update your local state with the new blocks AST
+  }
+});
+
+// CustomEvent dispatched by the preview script on the same window
+window.addEventListener('strapiFieldChange', (e) => {
```
Comment (giu1io) [inline docs/docs/docs/01-core/content-manager/06-preview.md]: **This makes an internal event part of the frontend API.**

Blocks live updates require host apps to listen for `strapiFieldChange`, but that event lives in `INTERNAL_EVENTS` — and line 94 of this same doc says internal events "can be changed freely since we control both ends." Once user frontends subscribe to it, we no longer control both ends.

The `CustomEvent` here is also newly introduced by this PR purely for host consumption, so it's a new external dependency on an explicitly-unstable name. Existing frontends won't live-update until they add this integration either.

Please expose a stable public event contract, or keep the integration inside the injected script so user frontends don't depend on an internal event name.

_Generated by Claude, edited by me._

File: examples/getstarted/src/admin/preview/dummy-preview.jsx [hunk 8]
```diff
@@ -63,8 +63,19 @@ const isAudioMedia = (value) => {
...
+    (node) =>
+      node !== null &&
+      typeof node === 'object' &&
+      typeof node.type === 'string' &&
+      Array.isArray(node.children)
+  );
+
 const NestedValue = ({ value, level = 0, arrayIndex = undefined, fieldName = undefined }) => {
-  if (fieldName === 'blocks') {
+  if (fieldName === 'blocks' || isBlocksValue(value)) {
```
Comment (giu1io) [inline examples/getstarted/src/admin/preview/dummy-preview.jsx:78]: **The getstarted preview still doesn't apply live blocks changes.**

This renders blocks with `BlocksRenderer`, but `PreviewComponent.handleMessage` (`:246`) only handles `strapiUpdate` and `strapiScript` — it never listens for `strapiFieldChange`, on `message` or as a `CustomEvent`. So it keeps rendering the loader's original value and typing in the popover can't update this preview in real time.

That makes step 4 of the PR's own test plan ("Type something — the preview should update in real time") unverifiable in `examples/getstarted`.

Please wire the event into local preview state so the sandbox exercises the complete workflow.

_Generated by Claude, edited by me._

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js [hunk 9]
```diff
@@ -505,10 +577,148 @@ function previewScript(config) {
...
+    /**
+     * Find the nearest common ancestor of a set of elements — the deepest
+     * single element that is an ancestor of (or equal to) every element in
+     * the list. Reliable for 2+ elements; for a single element it returns
+     * the immediate parent (which may not be the container you want — use
+     * the BLOCK_LEVEL_TAGS fallback in findBlocksFieldContainer instead).
+     * @param {HTMLElement[]} elements
+     * @returns {HTMLElement | null}
+     */
+    const findNearestCommonAncestor = (elements) => {
```
Comment (giu1io) [inline packages/core/content-manager/server/src/preview/controllers/previewScript.js]: **`findNearestCommonAncestor` is never called.**

The last commit replaced the NCA approach with the `BLOCK_LEVEL_TAGS` walk-up in both `findBlocksContainer` and `findBlockIndex`. This helper is now unreachable — nothing references it, and ESLint doesn't flag it because it sits inside the serialised script function.

It also ships to every preview iframe, since this whole function is stringified and injected.

Please delete it.

_Generated by Claude, edited by me._

File: docs/docs/docs/01-core/content-manager/06-preview.md [hunk 10]
```diff
@@ -96,3 +100,78 @@ Users can configure the preview behavior from their frontend via `window` global
...
+
+1. Walk up the DOM from any span in the group looking for an element whose direct children include a block-level tag (`P`, `H1`–`H6`, `UL`, `OL`, `BLOCKQUOTE`, `PRE`). That element is the field container.
+2. Use the container's `getBoundingClientRect()` as the highlight rect. This includes container padding and empty trailing blocks that have no stega spans.
+3. Fallback (if no container is found): union of all span rects plus an 80 px bottom buffer so that empty trailing blocks remain clickable.
+
+### Double-click flow
+
+Double-clicking a blocks field highlight opens the editor at the clicked block.
+
+1. The preview script calls `findBlockIndex(anchor, group)` to identify which block was clicked. It finds the field container (NCA for 2+ spans, walk-up fallback for single-span groups), then walks up from the clicked span until its parent is the container, and returns the DOM child index of that element.
```
Comment (giu1io) [inline docs/docs/docs/01-core/content-manager/06-preview.md]: **This step documents the NCA approach that was just removed.**

It says the script calls `findBlockIndex(anchor, group)` and finds the container via "NCA for 2+ spans, walk-up fallback for single-span groups". The actual code is `findBlockIndex(anchor)` — one argument, no `group` — and it uses the block-level-tag walk-up unconditionally. The in-code comment even explains why NCA was dropped (it landed on `<main>`/`<article>`).

The "Highlight grouping" section above has the same problem, crediting NCA for `findBlocksContainer`.

Please update both to match the walk-up implementation.

_Generated by Claude, edited by me._

File: packages/core/content-manager/admin/src/preview/utils/constants.ts [hunk 11]
```diff
@@ -1,6 +1,11 @@
 import { NotificationConfig } from '@strapi/admin/strapi-admin';
 import { MessageDescriptor } from 'react-intl';
 
+export const PREVIEW_HIGHLIGHT_COLORS = {
```
Comment (giu1io) [inline packages/core/content-manager/admin/src/preview/utils/constants.ts]: **This constant has no consumer.**

Nothing in `packages/*/src` reads `PREVIEW_HIGHLIGHT_COLORS`. `Preview.tsx:187` builds the colours from theme tokens instead:

```ts
highlightHoverColor: theme.colors.primary500,
highlightActiveColor: theme.colors.primary600,
```

The hardcoded values do match the light theme — `primary500` is `#7b79ff` and `primary600` is `#4945ff`. The divergence is in dark and custom themes: the dark theme effectively reverses the two (`primary500: #4945ff`, `primary600: #7b79ff`), so these constants would invert hover and active there.

Please delete it.

_Generated by Claude, edited by me._

**Review by giu1io — COMMENTED** [review]
Nice job!  💪 Only one blocker, the other ones are minor stuff.

File: packages/core/core/src/services/content-source-maps.ts (hunk 5 above)
Comment (mathildeleg) [reply packages/core/core/src/services/content-source-maps.ts]: Fixed — the image url is no longer encoded. Instead, we encode `alternativeText` only (when non-empty), which goes into the alt attribute of the rendered ``. The previewScript now reads and decodes the alt attribute specifically for media elements. This avoids any corruption of the src attribute.

_AI-assisted comment_

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js (hunk 6 above)
Comment (mathildeleg) [reply packages/core/content-manager/server/src/preview/controllers/previewScript.js]: Fixed — the walk-up in both `findBlocksContainer` and `findBlockIndex` now skips UL and OL ancestors in addition to LI. Since `@strapi/blocks-react-renderer` places nested lists as direct children of the outer `<ul>/<ol>` (not inside a `<li>`), the outer list would otherwise satisfy the block-level check and be mistaken for the field container.
A regression test was added: paragraph followed by a nested list, asserting that blockIndex is the `<ul>`'s position in the blocks container (1), not the `<li>`'s position within the list (0).

_AI-assisted comment_

File: docs/docs/docs/01-core/content-manager/06-preview.md (hunk 7 above)
Comment (mathildeleg) [reply docs/docs/docs/01-core/content-manager/06-preview.md]: Fixed — `STRAPI_FIELD_CHANGE` is now in `PUBLIC_EVENTS` (not `INTERNAL_EVENTS`), with a doc comment explaining its stability contract. The docs and the protocol diagram were updated to label it as a public event.

_AI-assisted comment_

File: examples/getstarted/src/admin/preview/dummy-preview.jsx (hunk 8 above)
Comment (mathildeleg) [reply examples/getstarted/src/admin/preview/dummy-preview.jsx:78]: Fixed — `PreviewComponent` now listens for the `strapiFieldChange` CustomEvent on window and merges updated field values into a `liveFields` state map. Those overrides are spread over the loader data at render time `({ ...main, ...liveFields })`, so `BlocksRenderer` re-renders in real time while the popover is open. The `strapiUpdate` handler also resets `liveFields` to clear stale overrides after a save.

_AI-assisted comment_

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js (hunk 9 above)
Comment (mathildeleg) [reply packages/core/content-manager/server/src/preview/controllers/previewScript.js]: Fixed — the function has been deleted.

_AI-assisted comment_

File: docs/docs/docs/01-core/content-manager/06-preview.md (hunk 10 above)
Comment (mathildeleg) [reply docs/docs/docs/01-core/content-manager/06-preview.md]: Fixed — both the "Highlight grouping" and "Double-click flow" sections now describe the block-level-tag walk-up approach. `findBlockIndex` is documented with its actual single-argument signature, and there's no remaining mention of NCA or the two-path strategy.

_AI-assisted comment_

File: packages/core/content-manager/admin/src/preview/utils/constants.ts (hunk 11 above)
Comment (mathildeleg) [reply packages/core/content-manager/admin/src/preview/utils/constants.ts]: Fixed — the constant has been removed. Colors are derived from `theme.colors.primary500 / theme.colors.primary600` directly in `Preview.tsx`, which also correctly adapts to dark and custom themes.

_AI-assisted comment_

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js [hunk 12]
```diff
@@ -666,9 +1062,33 @@ function previewScript(config) {
...
+              }
+              el = el.parentElement;
+            }
+          }
+          rect = anchorRect;
+        } else {
+          rect = computeGroupRect(group);
+        }
         if (!rect) return;
+        const blockIndex = isBlocksField ? findBlockIndex(anchor) : -1;
```
Comment (innerdvations) [inline packages/core/content-manager/server/src/preview/controllers/previewScript.js]: **Change request — do not jump the cursor to a neighboring block.**

The purple overlay covers the **whole** blocks field, including pieces we never mark:
- code blocks (encoding skipped on purpose)
- images with empty/missing alt (we only encode alt when it is a non-empty string)
- padding / empty space inside the field

`pickElementAtPoint` only knows about **marked** nodes. When the click is on an unmarked image or code block, nothing contains that point, so it falls back to the **nearest marked sibling** (a paragraph, heading, etc.). That sibling is then passed into `findBlockIndex`, and the Slate cursor lands on the **wrong block**.

That is more confusing than opening the field with no specific block: the UI looks precise, then the user edits the wrong paragraph.

**Please keep encoding as-is** (do not stega-encode `src` or code text).

**Please change only the blocks double-click path:**
1. If the click is inside a **marked** block’s rect → keep `findBlockIndex` (current happy path).
2. If it is not → still open the popover (edit the field), but send `blockIndex: null` so the editor does not jump to a sibling.

Do not use the nearest-marked-element fallback to compute `blockIndex`. That fallback is fine for “which field?”, not for “which Slate block?”.

`InputPopover` already treats a missing index as “no cursor jump” (`autoFocus` when `blockIndex == null`).

**Tests to add** (against this file / jsdom):
- Paragraph + empty-alt image; double-click the image → `blockIndex` is `null`, not `0`.
- Code block + following paragraph; double-click the code → `blockIndex` is `null`, not the paragraph’s index.
- Double-click the second marked paragraph → still `1`.

_AI-assisted comment._

**Review by innerdvations — CHANGES_REQUESTED** [review]
One more targeting issue on blocks double-click: if the click is not on a marked block (code, image without alt, padding), we still open the editor but should not jump the cursor to a neighboring block. Details inline.

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js [hunk 13]
```diff
@@ -505,10 +602,265 @@ function previewScript(config) {
...
+     * instead picks up any size change inside it (new blocks, growing or
+     * shrinking text, removed blocks) without requiring each child to be
+     * individually tagged. computeGroupRect already reads the container's
+     * rect fresh on every call — this only makes sure something re-triggers
+     * that read while the user is typing, instead of waiting for the next
+     * explicit rescan.
+     * @type {Map<string, HTMLElement>}
+     */
+    const observedContainers = new Map();
+    const containerResizeObserver = new ResizeObserver(() => {
```
Comment (giu1io) [inline packages/core/content-manager/server/src/preview/controllers/previewScript.js:619]: **`containerResizeObserver` is never disconnected on cleanup.**

It's created inside `createHighlightManager` and not exposed, so `__strapi_previewCleanup` can't reach it — it disconnects `resizeObserver`, `highlightObserver` and `stegaObserver`, but not this one.

Verified in `examples/getstarted` on this commit: after `__strapi_previewCleanup()` the per-element observer reports `disconnected: true` while this one is still `false` and still watching the field `DIV`, with the overlay already removed. Re-running the script 5 times leaks +1 live observer per run (3→4→5→6→7), each pinning a DOM node. The preview re-runs the script on navigation and on `strapiScript`, so this grows over an editing session.

Please expose it and disconnect it in `createCleanupSystem`.

Side note: the new test's `MockResizeObserver` keys `observedCallbacks` by target in a module-level map shared across instances, so one observer's `unobserve` deletes another's entry — which is why the suite can't see this. A per-instance mock makes both observers visible.

_Generated by Claude, edited by me._

**Review by giu1io — COMMENTED** [review]
Looks good, quick cleanup issue to address before my final QA.

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js (hunk 12 above)
Comment (mathildeleg) [reply packages/core/content-manager/server/src/preview/controllers/previewScript.js]: Here's the fix and the added tests: https://github.com/strapi/strapi/pull/27274/changes/638b69e6024d241fdb25129737ee07bd04ce81f6 

Let me know if there's still something missing/not right!

File: packages/core/content-manager/server/src/preview/controllers/previewScript.js (hunk 13 above)
Comment (mathildeleg) [reply packages/core/content-manager/server/src/preview/controllers/previewScript.js:619]: Good catch, here's the fix: https://github.com/strapi/strapi/pull/27274/changes/9eb3f4e37a94199c1e4e577de9f02dd75a7e21f8

**Review by innerdvations — DISMISSED** [review]
Looks good to merge. The earlier unmarked-click cursor jump is fixed (exact hit + `isKnownUnencodedElement`; live unmarked paragraphs still resolve), image-only highlights stay on the field, `null`/`[]` clears go through `strapiFieldChange`, and the container resize observer disconnects on cleanup.

Optional nits, none blocking:

- **Docs walk-up:** highlight grouping still says the container walk skips only `LI`. `findBlocksContainer` also skips `UL`/`OL` (nested lists as direct children of the outer list). Please mention that so it matches the code.
- **Cursor wording:** the double-click flow docs say `Transforms.select` at `[targetIndex, 0]`. The implementation uses `Editor.start(editor, [targetIndex])`, which is the right call for lists/headings.
- **getstarted dummy:** `liveFields` is still a shallow spread (`{ ...main, ...liveFields }`), so dotted nested paths like `teamMembers.bio` do not update. Example-only; the shipped preview script is fine. A deep-set would make nested component/DZ blocks update live in the dummy too.

_AI-assisted comment._

**Comment by mathildeleg** [comment]
> Looks good to merge. The earlier unmarked-click cursor jump is fixed (exact hit + `isKnownUnencodedElement`; live unmarked paragraphs still resolve), image-only highlights stay on the field, `null`/`[]` clears go through `strapiFieldChange`, and the container resize observer disconnects on cleanup.
> 
> Optional nits, none blocking:
> 
> * **Docs walk-up:** highlight grouping still says the container walk skips only `LI`. `findBlocksContainer` also skips `UL`/`OL` (nested lists as direct children of the outer list). Please mention that so it matches the code.
> * **Cursor wording:** the double-click flow docs say `Transforms.select` at `[targetIndex, 0]`. The implementation uses `Editor.start(editor, [targetIndex])`, which is the right call for lists/headings.
> * **getstarted dummy:** `liveFields` is still a shallow spread (`{ ...main, ...liveFields }`), so dotted nested paths like `teamMembers.bio` do not update. Example-only; the shipped preview script is fine. A deep-set would make nested component/DZ blocks update live in the dummy too.
> 
> _AI-assisted comment._

Addressed here https://github.com/strapi/strapi/pull/27274/changes/2cf1b3f695752aa5f1db786bb69074585a3753b7

**Review by innerdvations — APPROVED** [review]
(still) LGTM!
